### ISSUE TITLE
Wanted: uri joining utility function in rust

### ISSUE THREAD

**Issue by ry** [denoland/deno#955]
To work towards #866, we need to have a utility function which can do URL path resolution (resolving ".." and "." properly)

We currently use [mozilla's url crate](https://github.com/servo/rust-url) for this, but in order to avoid having two different URL parsing crates, I would like to use Hyper's [http::uri crate](https://docs.rs/http/0.1.13/http/uri/index.html) instead. 

However, this crate has less functionality - in particular we are missing a "join" function. 

I've outlined a first attempt at this below... but it doesn't handle ".." and "." resolution. It needs tests.

```rust
fn uri_join(
  base: uri::Uri,
  path: &str,
) -> Result<uri::Uri, http::Error> {
  let base_parts = base.into_parts();
  let base_path_and_query = base_parts.path_and_query.unwrap();
  let base_path = base_path_and_query.path();
  let joined_path = String::from(base_path) + "/" + path;
  let joined_path_and_query =
    uri::PathAndQuery::from_shared(joined_path.into()).unwrap();
  let mut joined_parts = uri::Parts::default();
  joined_parts.scheme = base_parts.scheme;
  joined_parts.authority = base_parts.authority;
  joined_parts.path_and_query = Some(joined_path_and_query);

  uri::Uri::from_parts(joined_parts)
}
```

This could even be released as an independent crate - maybe using the name "uri_util" - sent upstream to [the http crate](https://github.com/hyperium/http) - or put into deno in src/uri_util.rs.

**Comment by cedric05** [comment]
can i take this up?

**Comment by romandev** [comment]
If no one is working on this, I'll take this issue.

**Comment by Mohammed90** [comment]
FWIW, there's a paper on how to handle .. and . properly. See [“Lexical File Names in Plan 9 or Getting Dot-Dot Right"](https://9p.io/sys/doc/lexnames.html). Go's implementation of [path cleansing](https://golang.org/pkg/path/#Clean)
 is based on the paper, and summarized as such:
```
1. Replace multiple slashes with a single slash.
2. Eliminate each . path name element (the current directory).
3. Eliminate each inner .. path name element (the parent directory)
   along with the non-.. element that precedes it.
4. Eliminate .. elements that begin a rooted path:
   that is, replace "/.." by "/" at the beginning of a path.```

**Comment by bartlomieju** [comment]
Landed in #1039
